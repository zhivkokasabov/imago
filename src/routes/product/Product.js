import React, { Fragment, useMemo, useState } from 'react';
import { useParams } from 'react-router-dom';
import {
  useQuery
} from 'react-query';

import { queryKey, get } from '../../queries/product';

import styles from './Product.module.scss';

import Image from '../../components/Image/Image';
import Spinner from '../../components/Spinner/Spinner';
import Divider from '../../components/Divider/Divider';
import Button from '../../components/Button/Button';
import Checkbox from '../../components/Checkbox/Checkbox';

export default function Product() {
  let { id } = useParams();

  const { data, isLoading } = useQuery({
    queryKey: [queryKey, id],
    queryFn: () => get(id)
  });

  const [selectedLicenseId, setSelectedLicenseId] = useState();
  const cheapestLicense = useMemo(() => data?.usagelicences.sort((a, b) => a.credits >= b.credits ? 1 : -1)[0], [data]);

  const onLicenseChange = (id) => {
    setSelectedLicenseId(id);
  };

  if (isLoading) {
    return (
      <div>
        <Spinner />
      </div>
    );
  }

  return (
    <div className={styles.product}>
      <section>
        <h1>
          {data.title}
        </h1>
        <p>
          {data.caption}
        </p>
        <div className={styles['product__image-container']}>
          <Image src={data.mediathumb} alt={data.title} omitDomain loading="eager" />
        </div>
      </section>
      <aside>
        <h2 className={styles['product--large']}>
          Choose your License
        </h2>
        <div className={`${styles.info} ${styles['product--grey']}`}>
          <div className={styles.info__header}>
            <span className={styles['product--large']}>Premium RF</span>
            <div>
              <span className={styles['product__sub-text']}>{'from '}</span>
              <span data-testid="cheapestLicense" className={styles['product--large']}>{`${cheapestLicense.credits} Credits`}</span>
            </div>
          </div>
          <Divider variant="light" />
          <div className={styles.license__list}>
            {
              data.usagelicences.map((license, index) => (
                <Fragment key={license.licenseid}>
                  <label data-testid="licenseLabel" htmlFor={license.licenseid} className={styles.license}>
                    <input
                      type="radio"
                      name="license"
                      id={license.licenseid}
                      value={license.licenseid}
                      onChange={() => onLicenseChange(license.licenseid)}
                    />
                    <span className={styles.license__info}>
                      <span className={styles['license__name-container']}>
                        <Checkbox className={styles.license__checkbox} active={selectedLicenseId === license.licenseid} />
                        <span className={styles.license__name}>{license.name}</span>
                      </span>
                      <span>{`${license.credits} Credits`}</span>
                    </span>
                    <span className={styles['product__sub-text']}>{license.languagePointer}</span>
                  </label>
                  {
                    index < data.usagelicences.length - 1
                      ? (
                        <Divider className={styles.license__divider} variant="light" />
                      )
                      : (null)
                  }
                </Fragment>
              ))
            }
          </div>
        </div>
        <div className={styles.product__actions}>
          <Button variant="outlined">Add to Basket</Button>
          <Button variant="contained">Buy Now</Button>
        </div>
        <Divider className={styles.product__divider} />
        <dl>
          <dt>CREDIT</dt>
          <dd>{data.creator}</dd>
          <dt>IMAGE ID</dt>
          <dd>{data.mediaid}</dd>
          <dt>MAX SIZE</dt>
          <dd>{`${data.width}x${data.height} PIXELS`}</dd>
          <dt>DATE</dt>
          <dd>{new Date(data.creationDate).toLocaleDateString()}</dd>
          <dt>INFO</dt>
          <dd>{data.caption}</dd>
        </dl>
      </aside>
    </div>
  );
}
