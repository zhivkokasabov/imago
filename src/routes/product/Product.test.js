import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import {
  QueryClient,
  QueryClientProvider,
} from 'react-query';

import Product from './Product';
import * as Router from 'react-router-dom';

process.env.API_URL = '';
process.env.CDN = '';

jest.mock('react-router-dom', () => {
  const actual = jest.requireActual('react-router-dom');

  return {
    ...actual,
    useParams: jest.fn()
  };
});

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      cacheTime: 0
    }
  }
});

describe('Product route', () => {
  const id = '1';
  const licenses = {
    caption: 'IMAGO Creative: Stadtlandschaften - Straßen von Berlin Low angle view of World Clock in Alexanderplatz square, Berlin City Copyright: PabloxCamacho B18116043',
    cliplength: 0,
    creationdate: '2021-03-15 00:00:00.000',
    creator: 'PhotoAlto',
    height: 5180,
    mediathumb: 'https://www.imago-images.com/bild/st/0115446019/s.jpg',
    title: 'IMAGO Creative: Stadtlandschaften - Straßen von Berlin Low angle view of World Clock in Alexanderpl',
    usagelicences: [
        {
          credits: 110,
          languagePointer: 'licence3',
          name: 'Web',
          licenseid: 3
        },
        {
          credits: 20,
          name: 'Full RF',
          licenseid: 4
        }
    ],
    width: 7695
  };
  const renderComponent = () => render(
    <QueryClientProvider client={queryClient}>
      <Router.MemoryRouter>
        <Product />
      </Router.MemoryRouter>
    </QueryClientProvider>
  );

  beforeEach(() => {
    global.fetch = jest.fn(() => Promise.resolve({
      ok: true,
      status: 200,
      json: () => Promise.resolve(licenses)
    }));

    jest.spyOn(Router, 'useParams').mockReturnValue({ id });
  });

  afterEach(jest.clearAllMocks);

  it('should render Product route', async () => {
    const { asFragment, queryByText } = renderComponent();

    await waitFor(() => expect(queryByText('Choose your License')).toBeInTheDocument());
    expect(asFragment()).toMatchSnapshot();
  });

  it('should send initial API request with the product id from the uri', async () => {
    renderComponent();

    await waitFor(() => expect(global.fetch).toHaveBeenCalled());
    expect(global.fetch).toHaveBeenCalledWith(`media/${id}`);
  });

  it('should show loading while the request is ongoing', async () => {
    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
    await waitFor(() => expect(global.fetch).toHaveBeenCalled());
  });

  it('should use the cheapest license for the info header', async () => {
    const { getByTestId, queryByText } = renderComponent();

    await waitFor(() => expect(queryByText('Choose your License')).toBeInTheDocument());
    expect(getByTestId('cheapestLicense')).toHaveTextContent('20 Credits');
  });

  it('should mark selected licens', async () => {
    const { asFragment, queryByText, getAllByTestId } = renderComponent();

    await waitFor(() => expect(queryByText('Choose your License')).toBeInTheDocument());
    
    fireEvent.click(getAllByTestId('licenseLabel')[1]);

    expect(asFragment()).toMatchSnapshot();
  });
});
