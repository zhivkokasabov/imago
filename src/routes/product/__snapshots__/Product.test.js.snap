// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product route should mark selected licens 1`] = `
<DocumentFragment>
  <div
    class="product"
  >
    <section>
      <h1>
        IMAGO Creative: Stadtlandschaften - Straßen von Berlin Low angle view of World Clock in Alexanderpl
      </h1>
      <p>
        IMAGO Creative: Stadtlandschaften - Straßen von Berlin Low angle view of World Clock in Alexanderplatz square, Berlin City Copyright: PabloxCamacho B18116043
      </p>
      <div
        class="product__image-container"
      >
        <img
          alt="IMAGO Creative: Stadtlandschaften - Straßen von Berlin Low angle view of World Clock in Alexanderpl"
          class="image"
          loading="eager"
          src="https://www.imago-images.com/bild/st/0115446019/s.jpg"
        />
      </div>
    </section>
    <aside>
      <h2
        class="product--large"
      >
        Choose your License
      </h2>
      <div
        class="info product--grey"
      >
        <div
          class="info__header"
        >
          <span
            class="product--large"
          >
            Premium RF
          </span>
          <div>
            <span
              class="product__sub-text"
            >
              from 
            </span>
            <span
              class="product--large"
              data-testid="cheapestLicense"
            >
              20 Credits
            </span>
          </div>
        </div>
        <div
          class="divider__light divider"
        />
        <div
          class="license__list"
        >
          <label
            class="license"
            data-testid="licenseLabel"
            for="4"
          >
            <input
              id="4"
              name="license"
              type="radio"
              value="4"
            />
            <span
              class="license__info"
            >
              <span
                class="license__name-container"
              >
                <span
                  class="checkbox license__checkbox"
                >
                  <span />
                </span>
                <span
                  class="license__name"
                >
                  Full RF
                </span>
              </span>
              <span>
                20 Credits
              </span>
            </span>
            <span
              class="product__sub-text"
            />
          </label>
          <div
            class="divider__light divider license__divider"
          />
          <label
            class="license"
            data-testid="licenseLabel"
            for="3"
          >
            <input
              id="3"
              name="license"
              type="radio"
              value="3"
            />
            <span
              class="license__info"
            >
              <span
                class="license__name-container"
              >
                <span
                  class="checkbox checkbox--active license__checkbox"
                >
                  <span />
                </span>
                <span
                  class="license__name"
                >
                  Web
                </span>
              </span>
              <span>
                110 Credits
              </span>
            </span>
            <span
              class="product__sub-text"
            >
              licence3
            </span>
          </label>
        </div>
      </div>
      <div
        class="product__actions"
      >
        <button
          class="button button--outlined"
        >
          Add to Basket
        </button>
        <button
          class="button button--contained"
        >
          Buy Now
        </button>
      </div>
      <div
        class="divider__dark divider product__divider"
      />
      <dl>
        <dt>
          CREDIT
        </dt>
        <dd>
          PhotoAlto
        </dd>
        <dt>
          IMAGE ID
        </dt>
        <dd />
        <dt>
          MAX SIZE
        </dt>
        <dd>
          7695x5180 PIXELS
        </dd>
        <dt>
          DATE
        </dt>
        <dd>
          Invalid Date
        </dd>
        <dt>
          INFO
        </dt>
        <dd>
          IMAGO Creative: Stadtlandschaften - Straßen von Berlin Low angle view of World Clock in Alexanderplatz square, Berlin City Copyright: PabloxCamacho B18116043
        </dd>
      </dl>
    </aside>
  </div>
</DocumentFragment>
`;

exports[`Product route should render Product route 1`] = `
<DocumentFragment>
  <div
    class="product"
  >
    <section>
      <h1>
        IMAGO Creative: Stadtlandschaften - Straßen von Berlin Low angle view of World Clock in Alexanderpl
      </h1>
      <p>
        IMAGO Creative: Stadtlandschaften - Straßen von Berlin Low angle view of World Clock in Alexanderplatz square, Berlin City Copyright: PabloxCamacho B18116043
      </p>
      <div
        class="product__image-container"
      >
        <img
          alt="IMAGO Creative: Stadtlandschaften - Straßen von Berlin Low angle view of World Clock in Alexanderpl"
          class="image"
          loading="eager"
          src="https://www.imago-images.com/bild/st/0115446019/s.jpg"
        />
      </div>
    </section>
    <aside>
      <h2
        class="product--large"
      >
        Choose your License
      </h2>
      <div
        class="info product--grey"
      >
        <div
          class="info__header"
        >
          <span
            class="product--large"
          >
            Premium RF
          </span>
          <div>
            <span
              class="product__sub-text"
            >
              from 
            </span>
            <span
              class="product--large"
              data-testid="cheapestLicense"
            >
              20 Credits
            </span>
          </div>
        </div>
        <div
          class="divider__light divider"
        />
        <div
          class="license__list"
        >
          <label
            class="license"
            data-testid="licenseLabel"
            for="4"
          >
            <input
              id="4"
              name="license"
              type="radio"
              value="4"
            />
            <span
              class="license__info"
            >
              <span
                class="license__name-container"
              >
                <span
                  class="checkbox license__checkbox"
                >
                  <span />
                </span>
                <span
                  class="license__name"
                >
                  Full RF
                </span>
              </span>
              <span>
                20 Credits
              </span>
            </span>
            <span
              class="product__sub-text"
            />
          </label>
          <div
            class="divider__light divider license__divider"
          />
          <label
            class="license"
            data-testid="licenseLabel"
            for="3"
          >
            <input
              id="3"
              name="license"
              type="radio"
              value="3"
            />
            <span
              class="license__info"
            >
              <span
                class="license__name-container"
              >
                <span
                  class="checkbox license__checkbox"
                >
                  <span />
                </span>
                <span
                  class="license__name"
                >
                  Web
                </span>
              </span>
              <span>
                110 Credits
              </span>
            </span>
            <span
              class="product__sub-text"
            >
              licence3
            </span>
          </label>
        </div>
      </div>
      <div
        class="product__actions"
      >
        <button
          class="button button--outlined"
        >
          Add to Basket
        </button>
        <button
          class="button button--contained"
        >
          Buy Now
        </button>
      </div>
      <div
        class="divider__dark divider product__divider"
      />
      <dl>
        <dt>
          CREDIT
        </dt>
        <dd>
          PhotoAlto
        </dd>
        <dt>
          IMAGE ID
        </dt>
        <dd />
        <dt>
          MAX SIZE
        </dt>
        <dd>
          7695x5180 PIXELS
        </dd>
        <dt>
          DATE
        </dt>
        <dd>
          Invalid Date
        </dd>
        <dt>
          INFO
        </dt>
        <dd>
          IMAGO Creative: Stadtlandschaften - Straßen von Berlin Low angle view of World Clock in Alexanderplatz square, Berlin City Copyright: PabloxCamacho B18116043
        </dd>
      </dl>
    </aside>
  </div>
</DocumentFragment>
`;

exports[`Product route should show loading while the request is ongoing 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
